### PR TITLE
cmake: add ctest support to setup hook

### PR DESCRIPTION
**Pull request by jtojnar**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by matthewbauer** [comment]
Is there any way to get CMake to generate Makefiles with a check target? That would be nice if possible.

**Comment by jtojnar** [comment]
Using `ctest` is probably a better idea, since it works across backends.

**Comment by matthewbauer** [comment]
We really only support one backend in makefile, right now. According to this "make test" should already work:

https://gitlab.kitware.com/cmake/community/wikis/doc/ctest/Testing-With-CTest#simple-testing

Do you have any specific packages where tests are not being run right now when they should be?

**Comment by Mic92** [comment]
We also use the ninja backend. It would be great if we could switch to it by default.

**Comment by matthewbauer** [comment]
I think it’s a mistake to use Ninja + CMake in Nixpkgs. At least we shouldn’t do it in the per-package basis that happens right now in a few places. It’s definitely okay to use Ninja when the package has a hand-written ninja file (or in Meson’s case where the developers stubbornly /refuse/ to write a Make backend). But in the CMake case we should do the same thing everywhere. I do think it would be a good idea to add a check phase for Ninja when we are forced to use it, though.

**Comment by Mic92** [comment]
Can you elaborate why ninja is a problem? Make was not really designed as a backend for generated Makefiles.

**Comment by matthewbauer** [comment]
Mostly for legacy reasons - Make has been around for long enough that it seems worth supporting. Ninja depends on the Python runtime without really offering any clear advantages to Make. The time difference appears to be almost negligible:

http://david.rothlis.net/ninja-benchmark/

Of course, it's still worth supporting as a setup hook. I just think we should use Makefiles when they're available to us for consistency.

**Comment by Mic92** [comment]
* Ninja does not depend on python. I have seem projects that ship a single static executable of it to bootstrap their project
* this benchmark looks rather synthetic to me, when I add `ninja` to neovim the build takes 22.943 instead of 26.484, this is a 15 % percent speedup over the whole build.

**Comment by matthewbauer** [comment]
I suppose it's not hurting anything too much. My main rationale is that you should use as few dependencies as possible. Make is an implicit dependency on anything that use stdenv. Might as well use that when it is available.

It would definitely be nice if we could do some bootstrapping and avoid the build-time python dependency. Basic shell scripts could probably work? A static executable doesn't seem very cross platform though.

Anyway, I've opened a separate PR that adds the checPhase in the ninja setup hook:

https://github.com/NixOS/nixpkgs/pull/50527

**Comment by jtojnar** [comment]
We could also set `checkTarget` to `test` in cmake setup hook, but `ctest` looks cleaner to me and is analogous to what meson does.

**Comment by matthewbauer** [comment]
> We could also set checkTarget to test in cmake setup hook, but ctest looks cleaner to me and is analogous to what meson does.

I didn't realize that meson already had a check phase! That does change things but I am still partial to doing this all in Ninja. In setup.sh, build, install, & check are all handled by the same program (make). Ninja is supposed to be a make replacement so it should also handle the same phases. CMake, Meson, gyp, configure, etc. are all "generators" so they should only override the configure phase. This should hopefully lead to less phases in total and a clearer separation between build "generators" like CMake & Meson from "runners" like Make & Ninja.

File: pkgs/development/tools/build-managers/cmake/setup-hook.sh [hunk 1]
```diff
@@ -91,6 +91,18 @@ if [ -z "$dontUseCmakeConfigure" -a -z "$configurePhase" ]; then
     configurePhase=cmakeConfigurePhase
 fi
 
+cmakeCheckPhase() {
+    runHook preCheck
+
+    ctest --extra-verbose
```
Comment (orivej) [inline pkgs/development/tools/build-managers/cmake/setup-hook.sh]: Maybe `ctest --output-on-failure`, without `--extra-verbose`?

**Comment by jtojnar** [comment]
Closing in favour of #50527

**Pull request closed without merging**